### PR TITLE
Introduce `tx_graph::Update` and simplify `TxGraph` update logic

### LINKED ISSUE #1543: Make chain-sources non-dependent on a specific `bdk_chain` version.
## Describe the problem

We may want to be able to iterate fast on `bdk_chain` without breaking `Wallet`. This means `Wallet` may be of an older version of `bdk_chain`. Therefore, `Wallet` (currently) will have to use older versions of chain sources (since our chain sources depend on `bdk_chain`).

It will be a nice property if projects that use BDK can stay put on a specific `bdk_chain` version, but still use the latest chain-source crate version.

## Additional context

**Why would we need to break `bdk_chain`?**

There are a couple a breaking changes that we simply do not have enough time to include in the `v1.0-beta` release.

1. #1510
    Get rid of `IndexedTxGraph` and just put the `Indexer` in `TxGraph`. This removes a bunch of overhead and is in general much cleaner.
2. #1515
    Removing the anchor trait, and representing the `Anchor` as a concrete type is much cleaner and a less error-prone API.

Both of these changes will break `ChangeSet`.

## Proposed solution

Introduce `bdk_core` crate that moves structures and types from `bdk_chain` that will mostly stay constant.

* Types: `BlockId`, `ConfirmationBlockTime`.
* Updates: `CheckPoint`, introduce `tx_graph::Update` (instead of using `TxGraph` to update `TxGraph`).
* Spk-client types: `FullScanRequest`, `FullScanResult`, `SyncRequest`, `SyncResult`.

Chain sources will only depend on `bdk_core`. Note that we need to introduce `tx_graph::Update` though.

## Potential downsides

`bdk_core` will still need to depend on `bitcoin`. Any major version updates to `bitcoin` will be a breaking change and we will need to introduce a new major version for `bdk_core`. To make these suggested changes useful, we need to lock down on specific major releases of `bitcoin`.

## Acknowledgement

Thank you to @LLFourn for bringing up this idea.

### LINKED ISSUE #1550: Should `bdk_electrum` `sync`/`full_scan` set `last_seen`?
Getting a balance from `TxGraph` will not return a balance for unconfirmed transactions immediately after `sync`/`full_scan` with `bdk_electrum`. This is because the `last_seen` for these transactions is missing, and has to be set separately from `sync`/`full_scan` with `TxGraph::update_last_seen_unconfirmed`. Would it make sense to set the `last_seen` for unconfirmed transactions during the `sync`/`full_scan` process?

### PR DESCRIPTION
**Pull request by evanlinjin**
Part of #1543
Closes #1550

### Description

Instead of updating a `TxGraph` with another `TxGraph` in `.apply_update()`, we introduce `tx_graph::Update`. `tx_graph::Update` is a simple data object. This is the first step of #1543. This also makes it slightly less expensive to create an update.

Additionally, we simplify the update logic of `TxGraph` by containing most of the update logic in `.insert_{}`-esc methods. `.apply_update` and `.apply_changeset` will call `.insert_{}` methods internally and is greatly simplified. Thus, we also get rid of `.determine_changeset`.

We change `.apply_update` methods of `TxGraph`,`IndexedTxGraph` and `Wallet` to implicitly set `seen_at` to the current timestamp for unconfirmed update transactions. This makes these methods only available when the `"std"` cargo feature is enabled. For a non-std environment, we introduce `.apply_update_at` (as shown below).

```rust
pub fn apply_update_at(&mut self, update: Update<A>, seen_at: Option<u64>) -> ChangeSet<A> {
    todo!()
}
```

During this process, I've fixed some tests that didn't make much sense.

### Notes to the reviewers

There is a slight scope-creep, but they all address the update logic of `TxGraph` and structures that build on top.

### Changelog notice

* Changed update API (`.apply_update`) of `TxGraph` to take in a simple data object (`tx_graph::Update`).
* Changed `.apply_update` methods of `TxGraph`, `IndexedTxGraph` and `Wallet` to implicitly set the `seen_at` to the current timestamp for unconfirmed update transactions (this making them depend on the "std" cargo feature).
* Add `.apply_update_at` which is the no-std version of `.apply_update`.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

### CODE REVIEW

File: crates/chain/src/tx_graph.rs [hunk 1]
```diff
@@ -640,125 +691,70 @@ impl<A: Clone + Ord> TxGraph<A> {
     ///
     /// The returned [`ChangeSet`] is the set difference between `update` and `self` (transactions that
     /// exist in `update` but not in `self`).
-    pub fn apply_update(&mut self, update: TxGraph<A>) -> ChangeSet<A> {
-        let changeset = self.determine_changeset(update);
-        self.apply_changeset(changeset.clone());
+    #[cfg(feature = "std")]
```
Comment (LLFourn) [inline crates/chain/src/tx_graph.rs:689]:
```suggestion
    #[cfg(feature = "std")]
    #[cfg_attr(docsrs, doc(cfg(feature = "std")))]
```

Explain what's going on here internally.

File: crates/chain/tests/test_tx_graph.rs [hunk 2]
```diff
@@ -385,7 +310,7 @@ fn insert_txout_does_not_displace_tx() {
     let _ = tx_graph.insert_txout(
         OutPoint {
             txid: tx.compute_txid(),
-            vout: 0,
+            vout: 1,
```
Comment (LLFourn) [inline crates/chain/tests/test_tx_graph.rs:313]: looks like a mistaken change.

File: crates/esplora/src/blocking_ext.rs [hunk 3]
```diff
@@ -339,59 +354,48 @@ fn fetch_txs_with_spks<I: IntoIterator<Item = ScriptBuf>>(
...
         let handles = txids
             .by_ref()
             .take(parallel_requests)
             .map(|txid| {
                 let client = client.clone();
-                let tx_already_exists = tx_graph.get_tx(txid).is_some();
                 std::thread::spawn(move || {
-                    if tx_already_exists {
-                        client
-                            .get_tx_status(&txid)
```
Comment (LLFourn) [inline crates/esplora/src/blocking_ext.rs:362]: It looks like we were calling `get_tx_status` if we already had it but this seems like a bug so you removed it?

File: crates/chain/tests/test_tx_graph.rs [hunk 4]
```diff
@@ -88,57 +88,21 @@ fn insert_txouts() {
...
+        let mut update = tx_graph::Update::default();
         for (outpoint, txout) in &update_ops {
-            // Insert partials transactions
-            assert_eq!(
-                graph.insert_txout(*outpoint, txout.clone()),
-                ChangeSet {
-                    txouts: [(*outpoint, txout.clone())].into(),
-                    ..Default::default()
-                }
-            );
```
Comment (LLFourn) [inline crates/chain/tests/test_tx_graph.rs:100]: note sure if it's worth it but you could keep all the code here the same and then just turn the `TxGraph` into an `Update` at the end. Might as well to test the `From` impl but up to you.

File: crates/esplora/src/blocking_ext.rs (hunk 3 above)
Comment (evanlinjin) [reply crates/esplora/src/blocking_ext.rs:362]: Yes. I wouldn't call it a bug. It's just that esplora always returns a tx with a status. If we already have a tx, that means we already have the status. No need to getting the status here again

File: crates/chain/tests/test_tx_graph.rs (hunk 2 above)
Comment (LLFourn) [reply crates/chain/tests/test_tx_graph.rs:313]: oh no it's right.

File: crates/chain/tests/test_tx_graph.rs (hunk 4 above)
Comment (evanlinjin) [reply crates/chain/tests/test_tx_graph.rs:100]: I think it's better to add separate tests for those!

Comment (evanlinjin) [reply crates/chain/tests/test_tx_graph.rs:100]: Tests added.

**Review by notmandatory — APPROVED** [review]
ACK ccb8c796b259ae348cb8b8ef35b8c60e5621ccb5

Looks good and I can see how this will help decouple the blockchain clients from the wallet once `tx_graph:Update` is moved to a new crate. 

I ran all the examples also as a little smoke test and they worked fine.

**Review by LLFourn — APPROVED** [review]
ACK ccb8c796b259ae348cb8b8ef35b8c60e5621ccb5

**Pull request merged**